art.hpp

```diff
@@ -52,21 +91,285 @@ class db final {
+    // the degenerate case where the tree is a single root leaf, then
+    // the stack contains just that leaf.
+    //
+    // The [key] is the [std::byte] along which the path descends from
+    // that [node_ptr].  The [key] has no meaning for a leaf.  The key
+    // byte may be used to reconstruct the full key (along with any
+    // prefix bytes in the nodes along the path).  The key byte is
+    // tracked to avoid having to search the keys of some node types
+    // (N48) when the [child_index] does not directly imply the key
+    // byte.
+    //
+    // The [child_index] is the [std::uint8_t] index position in the
+    // parent at which the [child_ptr] was found.  The [child_index]
+    // has no meaning for a leaf.  In the special case of N48, the
+    // [child_index] is the index into the [child_indexes[]].  For all
+    // other internal node types, the [child_index] is a direct index
+    // into the [children[]].  When finding the successor (or
+    // predecessor) the [child_index] needs to be interpreted
+    // according to the node type.  For N4 and N16, you just look at
+    // the next slot in the children[] to find the successor.  For
+    // N256, you look at the next non-null slot in the children[].
+    // N48 is the oddest of the node types.  For N48, you have to look
+    // at the child_indexes[], find the next mapped key value greater
+    // than the current one, and then look at its entry in the
+    // children[].
+    std::stack<stack_entry> stack_ {};
+
+    // A buffer into which visited keys are decoded and materialized
+    // by get_key().
+    //
+    // Note: The internal key is a sequence of unsigned bytes having
+    // the appropriate lexicographic ordering.  The internal key needs
+    // to be decoded to the external key.
+    //
+    // FIXME The current implementation stores the entire key in the
+    // leaf. This works fine for simple primitive keys.  However, it
+    // needs to be modified when the implementation is modified to
+    // support variable length keys. In that situation, the full
+    // internal key needs to be constructed using the [key] byte from
+    // the path stack plus the prefix bytes from the internal nodes
+    // along that path.
+    key key_ {};
+    
+  }; // class iterator
+  
+  //
+  // end of the iterator API, which is an internal API.
+  //
+  
+ public:
+  
+  ///
+  /// public scan API
+  ///
+
+  // Note: The scan() interface is public.  The iterator and the
+  // methods to obtain an iterator are protected (except for tests).
+  // This encapsulation makes it easier to define methods which
+  // operate on external keys (scan()) and those which operate on
+  // internal keys (seek() and the iterator). It also makes life
+  // easier for mutex_db since scan() can take the lock.
+
+  // Scan the tree, applying the caller's lambda to each visited leaf.
+  //
+  // @param fn A function f(unodb::visitor<unodb::db::iterator>&)
```

Comment:
Let's start doing Doxygen in `///` style

Reply:
I am happy to iterate on the code documentation.  Let's do a documentation build so we have something we can work against.  I will open an issue for that.  See https://github.com/laurynas-biveinis/unodb/issues/621